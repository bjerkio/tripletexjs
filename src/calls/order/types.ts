import { DefaultTripletexInputs } from '../../types';

export interface ListOrdersInput extends DefaultTripletexInputs {
  /**
   * List of IDs
   */
  id?: string[];
}

export interface CreateOrderLineInput {
  productId: number;
  orderId: number;
  inventoryId?: number;
  inventoryLocationId?: number;
  vatTypeId?: number;

  description?: string;
  count?: number;
  unitCostCurrency?: number;
  unitPriceExcludingVatCurrency?: number;
  markup?: number;
  discount?: number;
  unitPriceIncludingVatCurrency?: number;
  isSubscription?: boolean;
  subscriptionPeriodStart?: Date;
  subscriptionPeriodEnd?: Date;
  // orderGroup?: OrderGroup;
}

export interface CreateOrderInput {
  // TODO: Add attn
  // attn?: string;

  customerId: number;
  contactId?: number;
  ourContactId?: number;
  ourContactEmployeeId?: number;
  projectId?: number;
  departmentId?: number;
  currencyId?: number;
  deliveryAddressId?: number;

  receiverEmail?: string;

  overdueNoticeEmail?: string;

  number?: string;

  reference?: string;

  orderDate: Date;

  /**
   * Comment to be displayed in the invoice based on this order.
   * Can be also found in Invoice.invoiceComment on Invoice objects.
   */
  invoiceComment?: string;

  /**
   * Number of days/months in which invoices created from this order is due
   */
  invoicesDueIn?: number;

  /**
   * Set the time unit of invoicesDueIn. The special case RECURRING_DAY_OF_MONTH
   * enables the due date to be fixed to a specific day of the month,
   * in this case the fixed due date will automatically be set as standard
   * on all invoices created from this order.
   *
   * Note that when RECURRING_DAY_OF_MONTH is set, the due date will be set
   * to the last day of month if "31" is set in invoicesDueIn.
   */
  invoicesDueInType?: 'DAYS' | 'MONTHS' | 'RECURRING_DAY_OF_MONTH';

  /**
   * Show account statement - open posts on invoices created from this order
   */
  isShowOpenPostsOnInvoices?: boolean;

  /**
   * Denotes if this order is closed. A closed order can no longer
   * be invoiced unless it is opened again.
   */
  isClosed?: boolean;

  deliveryDate: Date;

  deliveryComment?: string;
  isPrioritizeAmountsIncludingVat?: boolean;
  orderLineSorting?: 'ID' | 'PRODUCT' | 'CUSTOM';

  /**
   * If true, the order is a subscription, which enables periodical
   * invoicing of order lines. First, create an order with isSubscription=true,
   * then approve it for subscription invoicing with the :approveSubscriptionInvoice method.
   */
  isSubscription?: boolean;

  /**
   * Number of months/years the subscription shall run
   */
  subscriptionDuration?: number;

  /**
   * The time unit of subscriptionDuration
   */
  subscriptionDurationType?: 'MONTHS' | 'YEAR';

  /**
   * Number of periods on each invoice
   */
  subscriptionPeriodsOnInvoice?: number;

  /**
   * Invoicing in advance/in arrears
   */
  subscriptionInvoicingTimeInAdvanceOrArrears?: 'ADVANCE' | 'ARREARS';

  /**
   * Number of days/months invoicing in advance/in arrears
   */
  subscriptionInvoicingTime?: number;

  /**
   * The time unit of subscriptionInvoicingTime
   */
  subscriptionInvoicingTimeType?: 'DAYS' | 'MONTHS';

  /**
   * Automatic invoicing. Starts when the subscription is approved
   */
  isSubscriptionAutoInvoicing?: boolean;
}
