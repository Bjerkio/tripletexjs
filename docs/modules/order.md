[tripletexjs](../README.md) › [Order](order.md)

# Namespace: Order

## Index

### Enumerations

* [invoicesDueInType](../enums/order.invoicesdueintype.md)
* [orderLineSorting](../enums/order.orderlinesorting.md)
* [subscriptionDurationType](../enums/order.subscriptiondurationtype.md)
* [subscriptionInvoicingTimeInAdvanceOrArrears](../enums/order.subscriptioninvoicingtimeinadvanceorarrears.md)
* [subscriptionInvoicingTimeType](../enums/order.subscriptioninvoicingtimetype.md)
* [subscriptionPeriodsOnInvoiceType](../enums/order.subscriptionperiodsoninvoicetype.md)

### Properties

* [attn](order.md#optional-attn)
* [changes](order.md#optional-readonly-changes)
* [contact](order.md#optional-contact)
* [currency](order.md#optional-currency)
* [customer](order.md#customer)
* [deliveryAddress](order.md#optional-deliveryaddress)
* [deliveryComment](order.md#optional-deliverycomment)
* [deliveryDate](order.md#deliverydate)
* [department](order.md#optional-department)
* [id](order.md#optional-id)
* [invoiceComment](order.md#optional-invoicecomment)
* [invoicesDueIn](order.md#optional-invoicesduein)
* [isClosed](order.md#optional-isclosed)
* [isPrioritizeAmountsIncludingVat](order.md#optional-isprioritizeamountsincludingvat)
* [isShowOpenPostsOnInvoices](order.md#optional-isshowopenpostsoninvoices)
* [isSubscription](order.md#optional-issubscription)
* [isSubscriptionAutoInvoicing](order.md#optional-issubscriptionautoinvoicing)
* [number](order.md#optional-number)
* [orderDate](order.md#orderdate)
* [orderLines](order.md#optional-orderlines)
* [ourContact](order.md#optional-ourcontact)
* [ourContactEmployee](order.md#optional-ourcontactemployee)
* [overdueNoticeEmail](order.md#optional-overduenoticeemail)
* [project](order.md#optional-project)
* [receiverEmail](order.md#optional-receiveremail)
* [reference](order.md#optional-reference)
* [subscriptionDuration](order.md#optional-subscriptionduration)
* [subscriptionInvoicingTime](order.md#optional-subscriptioninvoicingtime)
* [subscriptionPeriodsOnInvoice](order.md#optional-subscriptionperiodsoninvoice)
* [url](order.md#optional-readonly-url)
* [version](order.md#optional-version)

## Properties

### `Optional` attn

• **attn**? : *[Contact](../interfaces/contact.md)*

___

### `Optional` `Readonly` changes

• **changes**? : *Array‹[Change](change.md)›*

___

### `Optional` contact

• **contact**? : *[Contact](../interfaces/contact.md)*

___

### `Optional` currency

• **currency**? : *[Currency](../interfaces/currency.md)*

___

###  customer

• **customer**: *[Customer](customer.md)*

___

### `Optional` deliveryAddress

• **deliveryAddress**? : *[DeliveryAddress](../interfaces/deliveryaddress.md)*

Delivery address of this order. This can be a new or existing address
(useful to know, especially if the delivery is to a private person: if 'deliveryAddress.name' is set, we ignore the state of 'customer.id')

___

### `Optional` deliveryComment

• **deliveryComment**? : *undefined | string*

___

###  deliveryDate

• **deliveryDate**: *string*

___

### `Optional` department

• **department**? : *[Department](../interfaces/department.md)*

___

### `Optional` id

• **id**? : *undefined | number*

___

### `Optional` invoiceComment

• **invoiceComment**? : *undefined | string*

Comment to be displayed in the invoice based on this order. Can be also found in Invoice.invoiceComment on Invoice objects.

___

### `Optional` invoicesDueIn

• **invoicesDueIn**? : *undefined | number*

Number of days/months in which invoices created from this order is due

___

### `Optional` isClosed

• **isClosed**? : *undefined | false | true*

Denotes if this order is closed. A closed order can no longer be invoiced unless it is opened again.

___

### `Optional` isPrioritizeAmountsIncludingVat

• **isPrioritizeAmountsIncludingVat**? : *undefined | false | true*

___

### `Optional` isShowOpenPostsOnInvoices

• **isShowOpenPostsOnInvoices**? : *undefined | false | true*

Show account statement - open posts on invoices created from this order

___

### `Optional` isSubscription

• **isSubscription**? : *undefined | false | true*

If true, the order is a subscription, which enables periodical invoicing of order lines. First, create an order with isSubscription=true, then approve it for subscription invoicing with the :approveSubscriptionInvoice method.

___

### `Optional` isSubscriptionAutoInvoicing

• **isSubscriptionAutoInvoicing**? : *undefined | false | true*

Automatic invoicing. Starts when the subscription is approved

___

### `Optional` number

• **number**? : *undefined | string*

___

###  orderDate

• **orderDate**: *string*

___

### `Optional` orderLines

• **orderLines**? : *Array‹[OrderLine](../interfaces/orderline.md)›*

Order lines tied to the order. New OrderLines may be embedded here, in some endpoints.

___

### `Optional` ourContact

• **ourContact**? : *[Contact](../interfaces/contact.md)*

If the contact is not an employee

___

### `Optional` ourContactEmployee

• **ourContactEmployee**? : *[Employee](employee.md)*

If the contact is an employee

___

### `Optional` overdueNoticeEmail

• **overdueNoticeEmail**? : *undefined | string*

___

### `Optional` project

• **project**? : *[Project](project.md)*

___

### `Optional` receiverEmail

• **receiverEmail**? : *undefined | string*

___

### `Optional` reference

• **reference**? : *undefined | string*

___

### `Optional` subscriptionDuration

• **subscriptionDuration**? : *undefined | number*

Number of months/years the subscription shall run

___

### `Optional` subscriptionInvoicingTime

• **subscriptionInvoicingTime**? : *undefined | number*

Number of days/months invoicing in advance/in arrears

___

### `Optional` subscriptionPeriodsOnInvoice

• **subscriptionPeriodsOnInvoice**? : *undefined | number*

Number of periods on each invoice

___

### `Optional` `Readonly` url

• **url**? : *undefined | string*

___

### `Optional` version

• **version**? : *undefined | number*
