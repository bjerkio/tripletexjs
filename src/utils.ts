import * as rt from 'runtypes';

export function pickFromObject<
  T extends Record<string, unknown>,
  K extends keyof T,
>(subject: T, ...keys: K[]): Pick<T, K> {
  const pairs = keys
    .map(key => [key, subject[key]])
    .filter(([, val]) => val !== undefined)
    .map(([key, val]) => [key, val]);
  return Object.fromEntries(pairs);
}

export function withRuntype<T>(validator: rt.Runtype<T>) {
  return (data: unknown) => {
    return validator.check(data);
  };
}

export function multipleValuesEnvelope<T>(type: rt.Runtype<T>) {
  return rt
    .Record({
      from: rt.Number,
      count: rt.Number,
      versionDigest: rt.String.nullable().optional(),
      values: rt.Array(type).asReadonly(),
    })
    .asReadonly();
}

export function singleValueEnvelope<T>(type: rt.Runtype<T>) {
  return rt
    .Record({
      value: type,
    })
    .asReadonly();
}
