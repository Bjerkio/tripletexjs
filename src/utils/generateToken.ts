import { OpenAPI } from '../generated/core/OpenAPI';
import { TokenSessionService } from '../generated/services/TokenSessionService';
import { invariant } from './invariant';

/**
 * generateToken helps creating a token. It sets token for all requests.
 * 
 * @param consumerToken Token of the API consumer
 * @param employeeToken The employees token
 * @param expirationDate Expiration date for the combined token
 * @example
 * ```typescript
 * const expDate = new Date();
 * expDate.setDate(expDate.getDate() + 1);
 * await generateToken({
 *   consumerToken: '000-000-000',
 *   employeeToken: '100-100-100',
 *   expirationDate: expDate,
 * });
 * ```
 */
export async function generateToken({
  expirationDate,
  ...tokenSessionArgs
}: {
  consumerToken: string;
  employeeToken: string;
  expirationDate: Date;
}) {
  if (OpenAPI.TOKEN) {
    return false;
  }

  const expDate = expirationDate.toISOString().split('T')[0];

  const session = await TokenSessionService.tokenSessionCreateCreate({
    expirationDate: expDate,
    ...tokenSessionArgs,
  });

  invariant(session.value.token, 'Not able to get token');

  OpenAPI.TOKEN = session.value;
}

/**
 * 0 or blank means the company of the employee.
 * Any other value means accountant clients. Use /company/>withLoginAccess to get a list of those.
 * @param username defaults to '0'
 */
export async function setUsername(username: string | (() => Promise<string>)) {
  OpenAPI.USERNAME = username;
}