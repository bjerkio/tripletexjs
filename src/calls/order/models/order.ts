import * as rt from 'runtypes';
import { multipleValuesEnvelope, resourceRef, singleValueEnvelope } from '../../../utils';
import { orderLineRt } from './order-line';

const orderRt = rt.Record({
  id: rt.Number,
  version: rt.Number,
  customer: resourceRef,
  contact: resourceRef,
  attn: resourceRef,
  receiverEmail: rt.String,
  overdueNoticeEmail: rt.String,
  number: rt.String,
  reference: rt.String,
  ourContact: resourceRef,
  ourContactEmployee: resourceRef,
  department: resourceRef,
  orderDate: rt.String,
  project: resourceRef,
  invoiceComment: rt.String,
  currency: resourceRef,
  invoicesDueIn: rt.Number,
  invoicesDueInType: rt.Union(
    rt.Literal('DAYS'),
    rt.Literal('MONTHS'),
    rt.Literal('RECURRING_DAY_OF_MONTH'),
  ),
  isShowOpenPostsOnInvoices: rt.Boolean,
  isClosed: rt.Boolean,
  deliveryDate: rt.String,
  deliveryAddress: resourceRef,
  deliveryComment: rt.String,
  isPrioritizeAmountsIncludingVat: rt.Boolean,
  orderLineSorting: rt.Union(
    rt.Literal('ID'),
    rt.Literal('PRODUCT'),
    rt.Literal('CUSTOM'),
  ),
  orderLines: rt.Array(orderLineRt),
  isSubscription: rt.Boolean,
  subscriptionDuration: rt.Number,
  subscriptionDurationType: rt.Union(rt.Literal('MONTHS'), rt.Literal('YEAR')),
  subscriptionPeriodsOnInvoice: rt.Number,
  subscriptionInvoicingTimeInAdvanceOrArrears: rt.Union(
    rt.Literal('ADVANCE'),
    rt.Literal('ARREARS'),
  ),
  subscriptionInvoicingTime: rt.Number,
  subscriptionInvoicingTimeType: rt.Union(
    rt.Literal('DAYS'),
    rt.Literal('MONTHS'),
  ),
  isSubscriptionAutoInvoicing: rt.Boolean,
});

export type Order = rt.Static<typeof orderRt>;

export const listOrderResponseRt = multipleValuesEnvelope(orderRt);
export const createOrderResponseRt = singleValueEnvelope(orderRt);
