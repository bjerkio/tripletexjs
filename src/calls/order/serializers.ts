import omitEmpty from 'omit-empty';
import { JsonRoot } from 'typical-fetch/dist/types';
import { toString } from '../../utils';
import { CreateOrderInput, CreateOrderLineInput } from './types';

export function makeOrderLineInput(input: CreateOrderLineInput) {
  const {
    productId,
    orderId,
    inventoryId,
    inventoryLocationId,
    vatTypeId,
    ...orderLine
  } = input;

  return omitEmpty({
    ...orderLine,
    product: {
      id: productId,
    },
    order: {
      id: orderId,
    },
    inventory: inventoryId && {
      id: inventoryId,
    },
    inventoryLocation: inventoryLocationId && {
      id: inventoryLocationId,
    },
    vatType: vatTypeId && {
      id: vatTypeId,
    },
  }) as JsonRoot;
}

export function makeOrderInput(input: CreateOrderInput) {
  const {
    customerId,
    contactId,
    ourContactId,
    ourContactEmployeeId,
    projectId,
    departmentId,
    currencyId,
    deliveryAddressId,
    ...order
  } = input;

  return omitEmpty({
    ...order,
    orderDate: order.orderDate && toString(order.orderDate),
    deliveryDate: order.deliveryDate && toString(order.deliveryDate),
    // subscriptionPeriodStart:
    //   order.subscriptionPeriodStart && toString(order.subscriptionPeriodStart),
    // subscriptionPeriodEnd:
    //   order.subscriptionPeriodEnd && toString(order.subscriptionPeriodEnd),
    customer: customerId && {
      id: customerId,
    },
    contact: contactId && {
      id: contactId,
    },
    ourContact: ourContactId && {
      id: ourContactId,
    },
    ourContactEmployee: ourContactEmployeeId && {
      id: ourContactEmployeeId,
    },
    project: projectId && {
      id: projectId,
    },
    department: departmentId && {
      id: departmentId,
    },
    currency: currencyId && {
      id: currencyId,
    },
    deliveryAddress: deliveryAddressId && {
      id: deliveryAddressId,
    },
  }) as JsonRoot;
}
