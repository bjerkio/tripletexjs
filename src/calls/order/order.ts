import { serializeQuery, withRuntype } from '../../utils';
import { TripletexBase } from '../base';
import { createOrderResponseRt, listOrderResponseRt } from './models/order';
import {
  createOrderLineResponseRt,
  createOrderLinesResponseRt,
} from './models/order-line';
import { makeOrderInput, makeOrderLineInput } from './serializers';
import {
  CreateOrderInput,
  CreateOrderLineInput,
  ListOrdersInput,
} from './types';
export * from './types';
export * from './models/order';
export * from './models/order-line';

export class TripletexOrder extends TripletexBase {
  list(input?: ListOrdersInput) {
    const call = this.authenticatedCall() //
      .args<{
        input?: ListOrdersInput;
      }>()
      .path('/v2/order')
      .query(args => (args.input ? serializeQuery(args.input) : {}))
      .method('get')
      .parseJson(withRuntype(listOrderResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ input, sessionToken }));
  }

  create(input: CreateOrderInput) {
    const call = this.authenticatedCall() //
      .args<{
        input: CreateOrderInput;
      }>()
      .path('/v2/order')
      .body(({ input }) => {
        return makeOrderInput(input);
      })
      .method('post')
      .parseJson(withRuntype(createOrderResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ input, sessionToken }));
  }

  createOrderLine(input: CreateOrderLineInput) {
    const call = this.authenticatedCall() //
      .args<{
        input: CreateOrderLineInput;
      }>()
      .path('/v2/order/orderline')
      .body(({ input }) => {
        return makeOrderLineInput(input);
      })
      .method('post')
      .parseJson(withRuntype(createOrderLineResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ input, sessionToken }));
  }

  createOrderLines(input: CreateOrderLineInput[]) {
    const call = this.authenticatedCall() //
      .args<{
        input: CreateOrderLineInput[];
      }>()
      .path('/v2/order/orderline/list')
      .body(({ input }) => {
        return input.map(orderLine => makeOrderLineInput(orderLine));
      })
      .method('post')
      .parseJson(withRuntype(createOrderLinesResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ input, sessionToken }));
  }
}
