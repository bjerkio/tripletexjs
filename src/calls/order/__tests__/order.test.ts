import nock, { activate } from 'nock';
import { invariant } from 'ts-invariant';
import { parseRuntypeValidationError } from '../../../utils';
import { TripletexOrder } from '../order';

const baseUrl = process.env.BASE_URL || 'https://api.tripletex.io/';
const sessionToken = process.env.SESSION_TOKEN || 'a-session-token';

describe('order', () => {
  let client: TripletexOrder;

  beforeEach(() => {
    client = new TripletexOrder({ baseUrl }, sessionToken);
  });

  it.skip('should list activities', async () => {
    const activities = await client.list({});

    parseRuntypeValidationError(activities.error);

    invariant(activities.success);

    expect(activities.body.values).toMatchSnapshot();
  });

  it('create a order', async () => {
    nock('https://api.tripletex.io:443', { encodedQueryParams: true })
      .post('/v2/order', {
        orderDate: '2022-01-01',
        deliveryDate: '2022-01-01',
        customer: { id: 11289298 },
      })
      .reply(201, {
        value: {
          id: 1580602,
          version: 0,
          url: 'api.tripletex.io/v2/order/1580602',
          customer: {
            id: 11289298,
            url: 'api.tripletex.io/v2/customer/11289298',
          },
          contact: null,
          attn: null,
          receiverEmail: '',
          overdueNoticeEmail: '',
          number: '35630692',
          reference: '',
          department: null,
          orderDate: '2022-01-01',
          project: null,
          invoiceComment: '',
          currency: { id: 1, url: 'api.tripletex.io/v2/currency/1' },
          invoicesDueIn: 14,
          invoicesDueInType: 'DAYS',
          isShowOpenPostsOnInvoices: false,
          isClosed: false,
          deliveryDate: '2022-01-01',
          deliveryAddress: null,
          deliveryComment: '',
          isPrioritizeAmountsIncludingVat: false,
          orderLineSorting: 'PRODUCT',
          orderLines: [],
          isSubscription: false,
          subscriptionDuration: 0,
          subscriptionDurationType: 'MONTHS',
          subscriptionPeriodsOnInvoice: 0,
          subscriptionPeriodsOnInvoiceType: 'MONTHS',
          subscriptionInvoicingTimeInAdvanceOrArrears: 'ADVANCE',
          subscriptionInvoicingTime: 0,
          subscriptionInvoicingTimeType: 'MONTHS',
          isSubscriptionAutoInvoicing: false,
          preliminaryInvoice: {
            id: 1109033,
            url: 'api.tripletex.io/v2/invoice/1109033',
          },
          attachment: [],
          sendMethodDescription: 'Faktura mÃ¥ sendes manuelt.',
          invoiceOnAccountVatHigh: false,
          totalInvoicedOnAccountAmountAbsoluteCurrency: 0,
        },
      });
    const order = await client.create({
      customerId: 11289298,
      orderDate: new Date('2022-01-01'),
      deliveryDate: new Date('2022-01-01'),
    });

    parseRuntypeValidationError(order.error);
    invariant(order.success);

    expect(order.body.value).toMatchSnapshot();
  });

  it('create a order line', async () => {
    nock('https://api.tripletex.io:443', { encodedQueryParams: true })
      .post('/v2/order/orderline', {
        count: 1337,
        product: { id: 21558874 },
        order: { id: 1580602 },
      })
      .reply(201, {
        value: {
          id: 1192712,
          version: 1,
          url: 'api.tripletex.io/v2/order/orderline/1192712',
          product: {
            id: 21558874,
            url: 'api.tripletex.io/v2/product/21558874',
          },
          inventory: { id: 6311, url: 'api.tripletex.io/v2/inventory/6311' },
          description: '',
          count: 1337,
          unitCostCurrency: 0,
          unitPriceExcludingVatCurrency: 0,
          currency: { id: 1, url: 'api.tripletex.io/v2/currency/1' },
          markup: 0,
          discount: 0,
          vatType: { id: 3, url: 'api.tripletex.io/v2/ledger/vatType/3' },
          amountExcludingVatCurrency: 0,
          amountIncludingVatCurrency: 0,
          order: { id: 1580602, url: 'api.tripletex.io/v2/order/1580602' },
          unitPriceIncludingVatCurrency: 0,
          isSubscription: false,
          subscriptionPeriodStart: null,
          subscriptionPeriodEnd: null,
          orderGroup: null,
        },
      });
    const order = await client.createOrderLine({
      orderId: 1580602,
      productId: 21558874,
      count: 1337,
    });

    parseRuntypeValidationError(order.error);
    invariant(order.success);

    expect(order.body.value).toMatchSnapshot();
  });

  it('create multiple order lines', async () => {
    nock('https://api.tripletex.io:443', { encodedQueryParams: true })
      .post('/v2/order/orderline/list', [
        { count: 1337, product: { id: 21558874 }, order: { id: 1580602 } },
        { count: 1338, product: { id: 21558874 }, order: { id: 1580602 } },
      ])
      .reply(201, {
        fullResultSize: 0,
        from: 0,
        count: 2,
        versionDigest: null,
        values: [
          {
            id: 1192909,
            version: 1,
            url: 'api.tripletex.io/v2/order/orderline/1192909',
            product: {
              id: 21558874,
              url: 'api.tripletex.io/v2/product/21558874',
            },
            inventory: {
              id: 6311,
              url: 'api.tripletex.io/v2/inventory/6311',
            },
            description: '',
            count: 1337,
            unitCostCurrency: 0,
            unitPriceExcludingVatCurrency: 0,
            currency: { id: 1, url: 'api.tripletex.io/v2/currency/1' },
            markup: 0,
            discount: 0,
            vatType: { id: 3, url: 'api.tripletex.io/v2/ledger/vatType/3' },
            amountExcludingVatCurrency: 0,
            amountIncludingVatCurrency: 0,
            order: { id: 1580602, url: 'api.tripletex.io/v2/order/1580602' },
            unitPriceIncludingVatCurrency: 0,
            isSubscription: false,
            subscriptionPeriodStart: null,
            subscriptionPeriodEnd: null,
            orderGroup: null,
          },
          {
            id: 1192910,
            version: 1,
            url: 'api.tripletex.io/v2/order/orderline/1192910',
            product: {
              id: 21558874,
              url: 'api.tripletex.io/v2/product/21558874',
            },
            inventory: {
              id: 6311,
              url: 'api.tripletex.io/v2/inventory/6311',
            },
            description: '',
            count: 1338,
            unitCostCurrency: 0,
            unitPriceExcludingVatCurrency: 0,
            currency: { id: 1, url: 'api.tripletex.io/v2/currency/1' },
            markup: 0,
            discount: 0,
            vatType: { id: 3, url: 'api.tripletex.io/v2/ledger/vatType/3' },
            amountExcludingVatCurrency: 0,
            amountIncludingVatCurrency: 0,
            order: { id: 1580602, url: 'api.tripletex.io/v2/order/1580602' },
            unitPriceIncludingVatCurrency: 0,
            isSubscription: false,
            subscriptionPeriodStart: null,
            subscriptionPeriodEnd: null,
            orderGroup: null,
          },
        ],
      });
    const order = await client.createOrderLines([
      {
        orderId: 1580602,
        productId: 21558874,
        count: 1337,
      },
      {
        orderId: 1580602,
        productId: 21558874,
        count: 1338,
      },
    ]);

    parseRuntypeValidationError(order.error);
    invariant(order.success);

    expect(order.body.values).toMatchSnapshot();
  });
});
