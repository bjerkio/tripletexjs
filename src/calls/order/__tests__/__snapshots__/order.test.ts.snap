// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`order create a order 1`] = `
Object {
  "attachment": Array [],
  "attn": null,
  "contact": null,
  "currency": Object {
    "id": 1,
    "url": "api.tripletex.io/v2/currency/1",
  },
  "customer": Object {
    "id": 11289298,
    "url": "api.tripletex.io/v2/customer/11289298",
  },
  "deliveryAddress": null,
  "deliveryComment": "",
  "deliveryDate": "2022-01-01",
  "department": null,
  "id": 1580602,
  "invoiceComment": "",
  "invoiceOnAccountVatHigh": false,
  "invoicesDueIn": 14,
  "invoicesDueInType": "DAYS",
  "isClosed": false,
  "isPrioritizeAmountsIncludingVat": false,
  "isShowOpenPostsOnInvoices": false,
  "isSubscription": false,
  "isSubscriptionAutoInvoicing": false,
  "number": "35630692",
  "orderDate": "2022-01-01",
  "orderLineSorting": "PRODUCT",
  "orderLines": Array [],
  "overdueNoticeEmail": "",
  "preliminaryInvoice": Object {
    "id": 1109033,
    "url": "api.tripletex.io/v2/invoice/1109033",
  },
  "project": null,
  "receiverEmail": "",
  "reference": "",
  "sendMethodDescription": "Faktura m√• sendes manuelt.",
  "subscriptionDuration": 0,
  "subscriptionDurationType": "MONTHS",
  "subscriptionInvoicingTime": 0,
  "subscriptionInvoicingTimeInAdvanceOrArrears": "ADVANCE",
  "subscriptionInvoicingTimeType": "MONTHS",
  "subscriptionPeriodsOnInvoice": 0,
  "subscriptionPeriodsOnInvoiceType": "MONTHS",
  "totalInvoicedOnAccountAmountAbsoluteCurrency": 0,
  "url": "api.tripletex.io/v2/order/1580602",
  "version": 0,
}
`;

exports[`order create a order line 1`] = `
Object {
  "amountExcludingVatCurrency": 0,
  "amountIncludingVatCurrency": 0,
  "count": 1337,
  "currency": Object {
    "id": 1,
    "url": "api.tripletex.io/v2/currency/1",
  },
  "description": "",
  "discount": 0,
  "id": 1192712,
  "inventory": Object {
    "id": 6311,
    "url": "api.tripletex.io/v2/inventory/6311",
  },
  "isSubscription": false,
  "markup": 0,
  "order": Object {
    "id": 1580602,
    "url": "api.tripletex.io/v2/order/1580602",
  },
  "orderGroup": null,
  "product": Object {
    "id": 21558874,
    "url": "api.tripletex.io/v2/product/21558874",
  },
  "subscriptionPeriodEnd": null,
  "subscriptionPeriodStart": null,
  "unitCostCurrency": 0,
  "unitPriceExcludingVatCurrency": 0,
  "unitPriceIncludingVatCurrency": 0,
  "url": "api.tripletex.io/v2/order/orderline/1192712",
  "vatType": Object {
    "id": 3,
    "url": "api.tripletex.io/v2/ledger/vatType/3",
  },
  "version": 1,
}
`;

exports[`order create multiple order lines 1`] = `
Array [
  Object {
    "amountExcludingVatCurrency": 0,
    "amountIncludingVatCurrency": 0,
    "count": 1337,
    "currency": Object {
      "id": 1,
      "url": "api.tripletex.io/v2/currency/1",
    },
    "description": "",
    "discount": 0,
    "id": 1192909,
    "inventory": Object {
      "id": 6311,
      "url": "api.tripletex.io/v2/inventory/6311",
    },
    "isSubscription": false,
    "markup": 0,
    "order": Object {
      "id": 1580602,
      "url": "api.tripletex.io/v2/order/1580602",
    },
    "orderGroup": null,
    "product": Object {
      "id": 21558874,
      "url": "api.tripletex.io/v2/product/21558874",
    },
    "subscriptionPeriodEnd": null,
    "subscriptionPeriodStart": null,
    "unitCostCurrency": 0,
    "unitPriceExcludingVatCurrency": 0,
    "unitPriceIncludingVatCurrency": 0,
    "url": "api.tripletex.io/v2/order/orderline/1192909",
    "vatType": Object {
      "id": 3,
      "url": "api.tripletex.io/v2/ledger/vatType/3",
    },
    "version": 1,
  },
  Object {
    "amountExcludingVatCurrency": 0,
    "amountIncludingVatCurrency": 0,
    "count": 1338,
    "currency": Object {
      "id": 1,
      "url": "api.tripletex.io/v2/currency/1",
    },
    "description": "",
    "discount": 0,
    "id": 1192910,
    "inventory": Object {
      "id": 6311,
      "url": "api.tripletex.io/v2/inventory/6311",
    },
    "isSubscription": false,
    "markup": 0,
    "order": Object {
      "id": 1580602,
      "url": "api.tripletex.io/v2/order/1580602",
    },
    "orderGroup": null,
    "product": Object {
      "id": 21558874,
      "url": "api.tripletex.io/v2/product/21558874",
    },
    "subscriptionPeriodEnd": null,
    "subscriptionPeriodStart": null,
    "unitCostCurrency": 0,
    "unitPriceExcludingVatCurrency": 0,
    "unitPriceIncludingVatCurrency": 0,
    "url": "api.tripletex.io/v2/order/orderline/1192910",
    "vatType": Object {
      "id": 3,
      "url": "api.tripletex.io/v2/ledger/vatType/3",
    },
    "version": 1,
  },
]
`;
